### PR TITLE
Fix structs to handle interfaces and arrays of interfaces or structs

### PR DESCRIPTION
**Pull request by ZetaDaemon**

### CODE REVIEW

File: Structs/__init__.py [hunk 1]
```diff
@@ -186,6 +186,18 @@ def convert_fstruct(
             )
 
         return expected_type(fstruct)  # type: ignore
+    
+    def convert_arg(
+        arg: Any,
+        idx: int
+    ) -> Any:
+        if isinstance(arg, unrealsdk.FArray):
+            return [convert_arg(a, i) for i, a in enumerate(arg)]
```
Comment (apple1417) [inline Structs/__init__.py]: This isn't going to work, the index is the index of the field on the struct, array entries aren't in the struct.

Prefer passing defaults in explicitly?

File: Structs/__init__.py [hunk 2]
```diff
@@ -186,6 +186,18 @@ def convert_fstruct(
             )
 
         return expected_type(fstruct)  # type: ignore
+    
+    def convert_arg(
```
Comment (apple1417) [inline Structs/__init__.py]: Add a docstring

**Review by apple1417 — CHANGES_REQUESTED** [review]

File: Structs/__init__.py (hunk 2 above)
Comment (ZetaDaemon) [reply Structs/__init__.py]: Resolved in latest commit

File: Structs/__init__.py (hunk 1 above)
Comment (ZetaDaemon) [reply Structs/__init__.py]: Resolved in latest commit

File: Structs/__init__.py [hunk 3]
```diff
@@ -189,12 +189,25 @@ def convert_fstruct(
...
+        Returns:
+            The new argument.
+        """
         if isinstance(arg, unrealsdk.FArray):
-            return [convert_arg(a, i) for i, a in enumerate(arg)]
-        if isinstance(arg, unrealsdk.FStruct) and idx < len(fields):
-            return convert_fstruct(arg, fields[idx], type(defaults[idx]))
+            return [convert_arg(a, f"{field}[{idx}]") for idx, a in enumerate(arg)]
+        if isinstance(arg, unrealsdk.FStruct):
+            return convert_fstruct(arg, field, _all_structs[arg.structType.Name])
```
Comment (apple1417) [inline Structs/__init__.py]: Need to use `_get_struct_name`, sometimes we prepend to the base name. Kwargs have a different way of getting it, though they won't work with your nicer array error message format, which I think's good to keep. Should be able to use this method for both.

I don't think it's cleanly possible, but would be neat if the field error string (maybe rename it to `field_error` too) could concatenate across nested structs - `... to field 'OuterField[0].Inner.InnerInner'`

File: Structs/__init__.py [hunk 4]
```diff
@@ -211,7 +224,8 @@ def new(cls: Type[NamedTuple], *args: Any, **kwargs: Any) -> NamedTuple:
 
         # Expand any nested fstructs
         args = [
-            convert_arg(val, idx)
+            val if idx > len(fields) else
+            convert_arg(val, fields[idx])
```
Comment (apple1417) [inline Structs/__init__.py]: If there are more args than fields, it's an error anyway - might as well just let the implicit index error fire and keep our code simpler

**Review by apple1417 — CHANGES_REQUESTED** [review]
You should probably also add this handling to kwargs

File: Structs/__init__.py (hunk 3 above)
Comment (ZetaDaemon) [reply Structs/__init__.py]: renamed `field` to `field_error`, kwargs is now handled similar to args (while keeping check for key in fields. Could not find solution in the current state to keep the entire field chain, remains a todo for the future

File: Structs/__init__.py (hunk 4 above)
Comment (ZetaDaemon) [reply Structs/__init__.py]: makes sense, updated and changed

File: Structs/__init__.py [hunk 5]
```diff
@@ -224,19 +224,18 @@ def new(cls: Type[NamedTuple], *args: Any, **kwargs: Any) -> NamedTuple:
...
-            for key, val in kwargs.items()
-        }
+			key: (
+				val
+				# If we have an invalid keyword arg, let the base __new__ deal with it
+				if key not in fields else
+				convert_arg(val, key)
+			)
+			for key, val in kwargs.items()
+		}
```
Comment (apple1417) [inline Structs/__init__.py]: Careful about tabs

File: Structs/__init__.py [hunk 6]
```diff
@@ -224,19 +224,18 @@ def new(cls: Type[NamedTuple], *args: Any, **kwargs: Any) -> NamedTuple:
...
-                # If we have an invalid keyword arg, let the base __new__ deal with it
-                if isinstance(val, unrealsdk.FStruct) and key in fields else
-                val
-            )
-            for key, val in kwargs.items()
-        }
+			key: (
+				val
+				# If we have an invalid keyword arg, let the base __new__ deal with it
+				if key not in fields else
```
Comment (apple1417) [inline Structs/__init__.py]: If you're using both branches anyway, don't use negative logic

**Review by apple1417 — APPROVED** [review]
Confirm it still works then I'll merge.

**Comment by ZetaDaemon** [comment]
tested and is still working

**Pull request merged**